onnxruntime/core/providers/cuda/math/binary_elementwise_ops.cc

```diff
@@ -404,31 +404,70 @@ Status Greater<T>::ComputeInternal(OpKernelContext* context) const {
 
   const Tensor* input0 = context->Input<Tensor>(0);
   const Tensor* input1 = context->Input<Tensor>(1);
+  auto input0_shape = input0->Shape();
+  auto input1_shape = input1->Shape();
+  size_t input0_size = input0_shape.Size();
+  size_t input1_size = input1_shape.Size();
+
   TensorShape output_shape;
-  ORT_RETURN_IF_ERROR(ComputeOutputShape(name, input0->Shape(), input1->Shape(), output_shape));
+  ORT_RETURN_IF_ERROR(ComputeOutputShape(name, input0_shape, input1_shape, output_shape));
   size_t output_size = output_shape.Size();
   Tensor* output_tensor = context->Output(0, output_shape);
 
   BinaryElementwisePreparation prepare(this);
-  ORT_RETURN_IF_ERROR(BinaryElementwiseBroadcastPrepare(input0, input1, output_tensor, &prepare));
-
   IAllocatorUniquePtr<T> output_buffer = GetScratchBuffer<T>(output_size);
-  Impl_Greater<CudaT>(
-      prepare.output_rank_or_simple_broadcast,
-      prepare.lhs_padded_strides.GpuPtr(),
-      reinterpret_cast<const CudaT*>(prepare.lhs_tensor->template Data<T>()),
-      prepare.rhs_padded_strides.GpuPtr(),
-      reinterpret_cast<const CudaT*>(prepare.rhs_tensor->template Data<T>()),
-      prepare.fdm_output_strides.GpuPtr(),
-      prepare.fdm_H,
-      prepare.fdm_C,
-      reinterpret_cast<CudaT*>(output_buffer.get()),
-      output_size);
+  if (input0_size == output_size) {
+    auto left_input = input0_size >= input1_size ? input0 : input1;
+    auto right_input = input0_size >= input1_size ? input1 : input0;
+    ORT_RETURN_IF_ERROR(BinaryElementwiseBroadcastPrepare(left_input, right_input, output_tensor, &prepare));
```

Comment:
This special case seems should have been handled for all binary elementwise ops. Please try reuse code.

Reply:
The reason for having a special code path here is case like A[4, 2], B[2] => C[4, 2], it can be handled simply and correctly here. But for case like A[2], B[4, 2] => C[4 2], it won't work here, need to go the code path in else{}

---
In reply to: [326831427](https://github.com/microsoft/onnxruntime/pull/1886#discussion_r326831427) [](ancestors = 326831427)

Reply:
Isn't that a case all other binary elementwise ops should handle?

---
In reply to: [326836759](https://github.com/microsoft/onnxruntime/pull/1886#discussion_r326836759) [](ancestors = 326836759,326831427)

Reply:
that's true. I can't remember who did the 1st one of Geater/Less/Equal. It was not quite the right way. Then other ops later just copied from it. I'll try to make it clean.

---
In reply to: [326839400](https://github.com/microsoft/onnxruntime/pull/1886#discussion_r326839400) [](ancestors = 326839400,326836759,326831427)